### PR TITLE
[Merged by Bors] - feat: Triangle counting

### PR DESCRIPTION
**Pull request by YaelDillies**
Prove the triangle counting lemma. Definitions that are internal to the proof are made private.

---

- [x] depends on: #8903

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Combinatorics/SimpleGraph/Triangle/Counting.lean [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+
+open Finset Fintype
+open scoped BigOperators
+
+variable {α : Type*} (G G' : SimpleGraph α) [DecidableRel G.Adj] {ε : ℝ} {s t u : Finset α}
+
+namespace SimpleGraph
+
+/-- The pairs of vertices whose density is big. -/
+private noncomputable def badVertices (ε : ℝ) (s t : Finset α) :=
```
Comment (TFBloom) [inline Mathlib/Combinatorics/SimpleGraph/Triangle/Counting.lean:32]: In this name (and throughout this file) you seem to be using 'vertices' to refer to parts of the partition of the big graph G. I understand why conceptually, but given how you've defined reduced graphs maybe would be less confusing to refer to these as parts instead? e.g. call this definition badParts and so on.

**Review by TFBloom — CHANGES_REQUESTED** [review]

File: Mathlib/Combinatorics/SimpleGraph/Regularity/Uniform.lean [hunk 2]
```diff
@@ -271,4 +294,162 @@ theorem nonuniformWitness_mem_nonuniformWitnesses (h : ¬G.IsUniform ε s t) (ht
...
+
 end Finpartition
+
+/-! ### Reduced graph -/
+
+namespace SimpleGraph
+
+/-- The reduction of the graph `G` along partition `P` has edges between `ε`-uniform pairs of parts
+that have edge density at least `δ`. -/
+@[simps] def reduced (ε δ : 𝕜) : SimpleGraph α where
```
[the remark this answers is not in the document]
Comment (YaelDillies) [reply Mathlib/Combinatorics/SimpleGraph/Regularity/Uniform.lean]: Good point! Renamed to `regularityReduced`

File: Mathlib/Combinatorics/SimpleGraph/Triangle/Counting.lean (hunk 1 above)
Comment (YaelDillies) [reply Mathlib/Combinatorics/SimpleGraph/Triangle/Counting.lean]: I actually think I mean honest to god vertices here? I've updated the docstring to make things clearer.

**Review by TFBloom — COMMENTED** [review]
The mathematical content all looks good here; definitions and statements agree with the literature, and the main result is a very useful tool in combinatorics - from the mathematical point of view this is ready to be merged.

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**